roseWong/assignments/slidingwindow/lc209/lc209.py

```diff
@@ -1,11 +1,32 @@
 # Problem Statement #
 
-# Given an array of positive numbers and a positive number ‘S,’ find the length of the smallest contiguous subarray whose sum is greater than or equal to ‘S’. Return 0 if no such subarray exists.
+# Given an array of positive numbers and a positive number ‘S’,
+# find the length of the smallest contiguous subarray whose sum is greater than or equal to ‘S’.
+# Return 0 if no such subarray exists.
+
+# starting with the code from 
 
 import math
 
 
 def smallest_subarray_with_given_sum(s, arr):
+  minLength = math.inf
+  windowLength = 0
+  windowSum = 0
+  windowStart = 0
+  windowEnd = 0
+  for windowEnd in range(len(arr)):
+    windowSum = windowSum + arr[windowEnd]  # this adds the next element in the array
+    windowLength = windowLength + 1
+    # slide the window when we hit the required size of s
+    while windowSum >= s:
+      minLength = min(minLength, (windowEnd - windowStart + 1))
+      # remove the element that is being dropped from the next sum
+      windowSum = windowSum - arr[windowStart]
+      windowStart = windowStart + 1
+      if minLength == math.inf:
+        return 0
+  return minLength
```

Comment:
Great solution! To build your muscle memory, write out the solution twice to understand it.